### ISSUE TITLE
get_interfaces for cumulus

### ISSUE THREAD

**Issue by jmrenshaw** [napalm-automation/napalm-cumulus#3]
This one looks like it may be a fairly simple one to start, the cumulus NCLU can provide the output to the "net show interface all" command in JSON format using "net show interface all json".  Example below:

```
cumulus@cumulus:~$ net show interface all json
{
	"eth0": {
		"connector_type": null,
		"iface_obj": {
			"asic": null,
			"connector_type": 0,
			"counters": null,
			"description": "",
			"ip_addr_assign": 1,
			"ip_address": {
				"allentries": [
					"192.168.0.136/24"
				]
			},
			"ip_neighbor": {
				"allentries": {
					"192.168.0.115": {
						"mac": "52:54:00:f4:41:a1"
					}
				}
			},
			"linkstate": 2,
			"lldp": null,
			"mac": "52:54:00:32:a5:a0",
			"mtu": 1500,
			"name": "eth0",
			"native_vlan": [],
			"speed": 1000,
			"vlan": null,
			"vlan_filtering": false,
			"vlan_list": {}
		},
		"linkstate": "UP",
		"name": "eth0",
		"port_category": "Mgmt",
		"speed": "1G",
		"summary": [
			"IP: 192.168.0.136/24(DHCP)"
		]
	}
}
```

First crack at the mappings between the keys:

NAPALM get_interfaces key | Cumulus NCLU key
------------ | -------------
is_up (True/False) | linkstate (UP/DOWN)
is_enabled (True/False) | doesn't seem to be an equivalent, always True?
description (string) | iface_obj description
last_flapped (int in seconds) | can't see any counters, may have to get this from somewhere else
speed (int in Mbit) | iface_obj speed
mac_address (string) | iface_obj mac

**Comment by GGabriele** [comment]
Thanks @jmrenshaw, this is usefulu!

`is_enabled`: I think we can use the `is_up` state for this
`last_flapped`: `net show counters` doesn't seems to have such information. @haderp do you know if it is available somewhere else?

**Comment by dbarrosop** [comment]
Where is the documentation for the data returned? I am also of the opinion that linkstate might return more information than just up/down. Your output seems to show "2"

**Comment by jmrenshaw** [comment]
I can't find any documentation related to this specific output unfortunately, so we may need to rely on a bit of experimentation to figure out what the "2" option means unless anyone else can jump in and provide a link.  Just to clarify my understanding of what the keys mean for NAPALM, is_up displays the physical state of the interface and is_enabled displays the administrative state?

**Comment by jmrenshaw** [comment]
If I was going to make a start on coding this first step is to clone the develop branch?

**Comment by GGabriele** [comment]
@jmrenshaw correct. You can try this:

```
$ git clone https://github.com/napalm-automation/napalm-cumulus
$ cd napalm-cumulus
$ git checkout -b new_branch
```

**Comment by jmrenshaw** [comment]
Thanks @GGabriele I have made a start on this and got the interface description, speed and mac dict keys working.  Still need to do some digging into what the numbers mean in the output for link state and find where the last_flapped information might be stored.  Quick question, is there a preferred response when an interface doesn't have a speed, i.e. loopback interfaces.  Cumulus responds with speed none, is this ok to copy directly into the get_interfaces response?  The documentation suggests this should be an integer so maybe change none for 0?

**Comment by GGabriele** [comment]
Nice job @jmrenshaw 

You can use `-1` for missing integer fields

**Comment by jmrenshaw** [comment]
I've done a bit of testing and found that the linkstate key under iface_obj is set to 0 for admin down and 2 when the interface is up.  Unfortunately I can't think of an easy way to test admin up but link down as I am using the virtual edition.  If anyone with access to a physical version could confirm that it is set to 1 when it is admin up/physical down that would be great as it would confirm the way I have coded it.
After doing a bit of reading it appears that there is no easy way to get the time since the last interface flap and this is definitely not available as standard for NCLU so I have just set last_flapped to -1 as standard.  This could easily be changed if this information is available in future releases.
Think I'm ready to upload the code, I've committed locally, what's the best way to get it onto github for people to review @GGabriele ?

**Comment by haderp** [comment]
@jmrenshaw @GGabriele  
For interface flaps, leverage quagga from the bash prompt `vtysh -c "show interface"`
Of course it is entirely dependent on the daemon running, but I think the vast majority of use-cases will be L3. Probably just add a test for is quagga daemon up?

Cumulus is working on a `net show` wrapper.

**Comment by jmrenshaw** [comment]
Thanks for this @haderp this has the info we need, @GGabriele as the output is not JSON for this particular command what is the preferred method of extracting the data we need?  Would a textFSM template be the preferred method?  Sample output is below:

Interface eth0 is up, line protocol is up
  Link ups:       0    last: (never)
  Link downs:     0    last: (never)
  PTM status: disabled
  vrf: Default-IP-Routing-Table
  index 2 metric 0 mtu 1500 
  flags: <UP,BROADCAST,RUNNING,MULTICAST>
  Type: Ethernet
  HWaddr: 52:54:00:32:a5:a0
  inet 192.168.0.136/24 broadcast 192.168.0.255
  inet6 fe80::5054:ff:fe32:a5a0/64
  Interface Type Other
Interface lo is up, line protocol is up
  Link ups:       0    last: (never)
  Link downs:     0    last: (never)
  PTM status: disabled
  vrf: Default-IP-Routing-Table
  index 1 metric 0 mtu 65536 
  flags: <UP,LOOPBACK,RUNNING>
  Type: Loopback
  Interface Type Other
Interface swp1 is down
  Link ups:       0    last: (never)
  Link downs:     0    last: (never)
  PTM status: disabled
  vrf: Default-IP-Routing-Table
  index 3 metric 0 mtu 1500 
  flags: <BROADCAST,MULTICAST>
  Type: Ethernet
  HWaddr: 52:54:00:1f:5a:a7
  Interface Type Other

**Comment by jmrenshaw** [comment]
I've had a go at putting in the last_flapped information from the quagga command on my [fork](https://github.com/jmrenshaw/napalm-cumulus/tree/get_interfaces).  Should I update the pull request?

**Comment by jmrenshaw** [comment]
This was merged with PR #6 so closing this issue.
